domain/blockdevice/state/state.go

```diff
@@ -0,0 +1,401 @@
+		var err error
+		result, err = loadBlockDevices(ctx, tx, machineId)
+		return errors.Trace(err)
+	})
+	return result, errors.Trace(err)
+}
+
+func loadBlockDevices(ctx context.Context, tx *sqlair.TX, machineId string) ([]blockdevice.BlockDevice, error) {
+	credQuery := `
+SELECT bd.* AS &BlockDevice.*,
+       bdl.* AS &DeviceLink.*,
+       fs_type.* AS &FilesystemType.*
+FROM   block_device bd
+       JOIN block_device_machine bdm ON bd.uuid = bdm.block_device_uuid
+       JOIN machine ON bdm.machine_uuid = machine.uuid
+       LEFT JOIN block_device_link_device bdl ON bd.uuid = bdl.block_device_uuid
+       LEFT JOIN filesystem_type fs_type ON bd.filesystem_type_id = fs_type.id
+WHERE  machine.machine_id = $M.machine_id
+`
+
+	types := []any{
+		BlockDevice{},
+		FilesystemType{},
+		DeviceLink{},
+		sqlair.M{},
+	}
+
+	stmt, err := sqlair.Prepare(credQuery, types...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var (
+		dbRows            BlockDevices
+		dbDeviceLinks     []DeviceLink
+		dbFilesystemTypes []FilesystemType
+	)
+	machineParam := sqlair.M{"machine_id": machineId}
+	err = tx.Query(ctx, stmt, machineParam).GetAll(&dbRows, &dbDeviceLinks, &dbFilesystemTypes)
+	if err != nil {
+		return nil, errors.Annotatef(err, "loading block devices for machine %q", machineId)
+	}
+	return dbRows.toBlockDevices(dbDeviceLinks, dbFilesystemTypes)
+}
+
+// GetMachineInfo is used look up the machine UUID and life for a machine.
+func (st *State) GetMachineInfo(ctx context.Context, machineId string) (string, domain.Life, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", 0, errors.Trace(err)
+	}
+
+	var (
+		machineUUID string
+		life        domain.Life
+	)
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var err error
+		machineUUID, life, err = getMachineInfo(ctx, tx, machineId)
+		return errors.Trace(err)
+	})
+	return machineUUID, life, errors.Trace(err)
+}
+
+func getMachineInfo(ctx context.Context, tx *sqlair.TX, machineId string) (string, domain.Life, error) {
+	q := `
+SELECT machine.life_id AS &M.life_id, machine.uuid AS &M.machine_uuid
+FROM   machine
+WHERE  machine.machine_id = $M.machine_id
+`
+	stmt, err := sqlair.Prepare(q, sqlair.M{})
+	if err != nil {
+		return "", 0, errors.Trace(err)
+	}
+
+	result := sqlair.M{}
+	err = tx.Query(ctx, stmt, sqlair.M{"machine_id": machineId}).Get(result)
+	if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return "", 0, errors.Trace(err)
+	}
+	if len(result) == 0 {
+		return "", 0, fmt.Errorf("machine %q %w", machineId, errors.NotFound)
+	}
+	life, ok := result["life_id"].(int64)
+	if !ok {
+		return "", 0, errors.Errorf("missing life value for machine %q", machineId)
+	}
+	machineUUID := result["machine_uuid"].(string)
+	return machineUUID, domain.Life(life), nil
+}
+
+// SetMachineBlockDevices sets the block devices visible on the machine.
+// Previously recorded block devices not in the list will be removed.
```

Comment:
Assuming we take on the machine not found error. I think you should indicate this error in the contract as something that can be returned.